tests/test_js_spidermonkey.py

```diff
@@ -40,52 +37,84 @@ def test_scripting_snippet():
-        return MockSubprocessObject()
+    assert not result.failed()
 
+    eq_(_get_var(result, "foo"), u"\u263a''")
 
-class MockSubprocessObject(object):
-    "A class to mock a subprocess object and implement the communicate method"
 
-    def communicate(self, data):
-        data = json.dumps({"error": True,
-                           "error_message": "ReferenceError: Reflect is not defined",
-                           "line_number": 0})
-        return data, ""
+@mock.patch("subprocess.Popen")
+def test_cleanup(Popen):
+    """Test that the spidermonkey processes are terminated during cleanup."""
 
+    # Make sure we don't already have a cached shell.
+    JSShell.cleanup()
+    eq_(JSShell.shells, {})
+    assert not Popen.called
 
-@mock.patch("validator.testcases.javascript.spidermonkey.subprocess",
-            MockSubprocess())
-def test_reflectparse_presence():
-    "Tests that when Spidermonkey is too old, a proper error is produced"
+    process = mock.Mock()
+    process.stdout.readline.return_value = "{}\n"
+    Popen.return_value = process
 
-    try:
-        spidermonkey._get_tree("foo bar", "[path]")
-    except RuntimeError as err:
-        assert str(err).startswith("Spidermonkey version too old"), err
+    get_tree("hello();", shell=SPIDERMONKEY_INSTALLATION)
 
+    eq_(Popen.call_count, 1)
+    eq_(JSShell.shells.keys(), [SPIDERMONKEY_INSTALLATION])
+    assert not process.terminate.called
 
-def test_compiletime_errors():
-    "Tests that compile time errors don't break the validator"
+    decorator.cleanup()
 
-    # Syntax error
-    assert _do_test_raw("var x =;").failed()
+    assert process.terminate.called
+    eq_(JSShell.shells, {})
 
-    # Reference error
-    assert _do_test_raw("x - y = 4;").failed()
 
+@mock.patch("subprocess.Popen")
+def test_cleanup_on_error(Popen):
+    """Test that the cached shells are removed on IO error."""
 
-def test_unicode_escapes():
-    """Tests that unicode/hex escapes are passed through to Spidermonkey."""
+    # Make sure we don't already have a cached shell.
+    JSShell.cleanup()
+    eq_(JSShell.shells, {})
 
-    result = _do_test_raw("var foo = '\\u263a\\u0027\\x27'")
+    process = mock.Mock()
+    process.stdout.readline.return_value = "{}"
+    Popen.side_effect = lambda *args, **kw: process
 
-    assert not result.failed()
+    # Success, should have cached shell.
+    get_tree("hello();", shell=SPIDERMONKEY_INSTALLATION)
+    eq_(Popen.call_count, 1)
+
+    eq_(JSShell.shells.keys(), [SPIDERMONKEY_INSTALLATION])
+    assert not process.terminate.called
+
+    # Read error. No cached shell.
+    process.stdout.readline.side_effect = IOError
+
+    with assert_raises(IOError):
+        get_tree("hello();", shell=SPIDERMONKEY_INSTALLATION)
+
+    # No new processes created, since we had a cached shell from last time.
+    eq_(Popen.call_count, 1)
+    assert process.stdin.write.called
+    assert process.stdout.readline.called
```

Comment:
I originally had it, but the deconstructor doesn't get called immediately in this test, so it fails. There are other tests for the deconstructor, so it didn't seem worthwhile to try to find a way to make that test work here.